Add GPTQ Quantization (#1216)
huggingface/optimum: docs/source/optimization_toolbox/usage_guides/quantization.mdx

# Quantization

## `AutoGPTQ` Integration

> 🤗 Optimum collaborated with `AutoGPTQ` library to provides a simple API that perform GPTQ quantization on language models.  With GPTQ quantization, you can quantize your favorite lanmguage model to 8,6,4 or even 2 bits. This comes without a big drop of performance and with faster inference speed. This is supported by most GPU hardwares.

Reviewer: ```suggestion
🤗 Optimum collaborated with `AutoGPTQ` library to provide a simple API that apply GPTQ quantization on language models.  With GPTQ quantization, you can quantize your favorite lanmguage model to 8,6,4 or even 2 bits. This comes without a big drop of performance and with faster inference speed. This is supported by most GPU hardwares.
```

If you want to quantize 🤗 Transformers models with GPTQ, follow this [documentation](https://huggingface.co/docs/transformers/main_classes/quantization). 

To learn more about the the quantization model, check out: 
- the [GPTQ](https://arxiv.org/pdf/2210.17323.pdf) paper
- the [`AutoGPTQ`](https://github.com/PanQiWei/AutoGPTQ) library used as the backend

> 

Reviewer: ```suggestion
For more advanced usage (triton backend and fused attention), please check out the `auto-gptq` library in more details.

```

### Requirements

You need to have the following requirements installed to run the code below: 

- Install latest `AutoGPTQ` library
`pip install auto-gptq`

> - Install latest `optimum` from source 
> `pip install --upgrade optimum`
> 
> - Install latest `transformers` from source 
> `pip install --upgrade transformers`
> 
> - Install latest `accelerate` from source 
> `pip install --upgrade accelerate`

Reviewer: ```suggestion
- Install latest `optimum` from source 
`pip install --upgrade git+https://github.com/huggingface/optimum.git`

- Install latest `transformers` from source 
`pip install --upgrade git+https://github.com/huggingface/transformers.git`

- Install latest `accelerate` library 
`pip install --upgrade accelerate`
```

# Quantization

## `AutoGPTQ` Integration

🤗 Optimum collaborated with `AutoGPTQ` library to provide a simple API that apply GPTQ quantization on language models.  With GPTQ quantization, you can quantize your favorite lanmguage model to 8,6,4 or even 2 bits. This comes without a big drop of performance and with faster inference speed. This is supported by most GPU hardwares.

If you want to quantize 🤗 Transformers models with GPTQ, follow this [documentation](https://huggingface.co/docs/transformers/main_classes/quantization). 

To learn more about the the quantization model, check out: 
- the [GPTQ](https://arxiv.org/pdf/2210.17323.pdf) paper
- the [`AutoGPTQ`](https://github.com/PanQiWei/AutoGPTQ) library used as the backend
For more advanced usage (triton backend and fused attention), please check out the `auto-gptq` library in more details.

### Requirements

You need to have the following requirements installed to run the code below: 

- Install latest `AutoGPTQ` library
`pip install auto-gptq`

- Install latest `optimum` from source 

> `pip install --upgrade git+https://github.com/huggingface/optimum.git`

Author: sounds good

- Install latest `transformers` from source 

> `pip install --upgrade git+https://github.com/huggingface/transformers.git`

Reviewer 2: Same, not sure we want that. Also, why would 4.31 not work?

Author: indeed, any version of transformers should work.

Reviewer 2: Adding the `QuantizationMethod.GPTQ`, I think this won't work anymore with 4.31.

Author: nice catch

- Install latest `accelerate` library 
`pip install --upgrade accelerate`

> ### Load and quantize model

Reviewer 2: ```suggestion
### Load and quantize a model
```

The [`~optimum.gptq.GPTQQuantizer`] class is used to quantize your model. In order to quantize your model, you need to provide a few arguemnts:
- the number of bits: `bits`
- the dataset used to calibrate the quantization: `dataset`
- the model sequence length used to process the dataset: `model_seqlen`
- the block name to quantize: `block_name_to_quantize`

With 🤗 Transformers integration, you don't need to pass the `block_name_to_quantize` and `model_seqlen` as we can retrieve them. However, for custom model, you need to specify them. Also, make sure that your model is converted to `torch.float16` before quantization.

```py

> from transformers import AutoModelForCausalLM, AutoTokenizer

Author: fixed

from optimum.gptq import GPTQQuantizer, load_quantized_model
model_name = "facebook/opt-125m"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16)

quantizer = GPTQQuantizer(bits=4, dataset="c4", block_name_to_quantize = "model.decoder.layers", model_seqlen = 2048)

> quantized_model = quantizer.quantize_model(quantized_model, tokenizer)

Reviewer 2: `quantized_model` is not defined here

```

<Tip warning={true}>
GPTQ quantization only works for text model for now. Futhermore, the quantization process can a lot of time depending on one's hardware (175B model = 4 gpu hours using NVIDIA A100). Please check on the hub if there is not a GPTQ quantized version of the model. If not, you can submit a demand on github. 
</Tip>

### Save model

To save your model, use the save method from [`~optimum.gptq.GPTQQuantizer`] class. It will create a folder with your model state dict along with the quantization config.
```python
save_folder = "/path/to/save_folder/"
quantizer.save(model,save_folder)
```

### Load quantized weights

You can load your quantized weights by using the [`~optimum.gptq.load_quantized_model`] function.
```python
from optimum.gptq import load_quantized_model

> model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16)

Reviewer 2: `model_name` is not defined, it would help if example codes are copy pastable and runable

Author: ofc

> model_from_saved = load_quantized_model(model, save_folder=save_folder, device_map = "auto")

Reviewer 2: would
```suggestion
quantized_model = load_quantized_model(model, save_folder=save_folder, device_map = "auto")
```
make sense? Does it duplicate memory between model + quantized_model?

> model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16)
> model_from_saved = load_quantized_model(model, save_folder=save_folder, device_map = "auto")

Reviewer 2: Would
```suggestion
with torch.device("meta"):
    model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16)
quantized_model = load_quantized_model(model, save_folder=save_folder, device_map = "auto")
```

work?

Author: It will indeed work. `init_empty_weights` from accelerate do that but it can also take care of buffer if specified. I forgot that load_quantized_model relies on accelerate. I think it will be better to merge the two examples.

quantized_model = load_quantized_model(empty_model, save_folder=save_folder, device_map="auto")
```

### Exllama kernels for faster inference

> For 4-bit model, you can use the exllama kernels in order to a faster inference speed. You just need to pass `disable_exllama = False` in [`~optimum.gptq.load_quantized_model`]. In order to use these kernels, you need to have the entire model on gpus.

Reviewer 2: Isn't False the default?

Author: Yes, the default behavior is [False](url). I will clarify the message.
